### PR TITLE
US012 add new role

### PR DESCRIPTION
**Pull request by joshuasy10**

### CODE REVIEW

File: src/main/java/org/kainos/ea/controller/BandController.java
Comment (jedgar3) [inline src/main/java/org/kainos/ea/controller/BandController.java:1]: Lokking good!

File: src/main/java/org/kainos/ea/controller/CapabilityController.java
Comment (jedgar3) [inline src/main/java/org/kainos/ea/controller/CapabilityController.java:1]: Good job

File: src/main/java/org/kainos/ea/dao/BandDao.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                    rs.getInt("id"),
+                    rs.getString("name"),
+                    rs.getString("level"),
+                    rs.getString("responsibilities")
+            );
+        }
+
+        return null;
+    }
+
```
Comment (jedgar3) [inline src/main/java/org/kainos/ea/dao/BandDao.java:59]: Remove whitespace

File: src/main/java/org/kainos/ea/model/JobRoleRequest.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                          @JsonProperty("bandId") int bandId,
+                          @JsonProperty("capabilityId") int capabilityId,
+                          @JsonProperty("specification") String specification) {
+        setName(name);
+        setBandId(bandId);
+        setCapabilityId(capabilityId);
+        setSpecification(specification);
+    }
+
+    public JobRoleRequest(String name, int band, int capability) {
```
Comment (jedgar3) [inline src/main/java/org/kainos/ea/model/JobRoleRequest.java:22]: Does specification need to be passed in as it is being set in this constructor. Is the constructor needed if it has the same parameters as the one above

File: src/main/java/org/kainos/ea/validator/JobRoleValidator.java
Comment (jedgar3) [inline src/main/java/org/kainos/ea/validator/JobRoleValidator.java:1]: Good validation work Josh

File: src/test/java/integration/controller/JobRoleControllerIntegrationTest.java [hunk 3]
```diff
@@ -48,4 +55,13 @@ void getJobRoleById_shouldReturn400_whenIDNotFound() {
 
         Assertions.assertEquals(400, responseJobRole.getStatus());
     }
-}
+
+    @Test
+    void createJobRole_shouldReturn201_whenCreated() {
```
Comment (jedgar3) [inline src/test/java/integration/controller/JobRoleControllerIntegrationTest.java:60]: should this method name say should return 200

File: src/main/java/org/kainos/ea/validator/JobRoleValidator.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        if (jobRoleRequest.getBandId() <= 0) {
+            throw new InvalidJobRoleException("Band ID cannot be 0 or less");
+        }
+        if (jobRoleRequest.getCapabilityId() <= 0) {
+            throw new InvalidJobRoleException("Capability ID cannot be 0 or less");
+        }
+        if (jobRoleRequest.getSpecification() == null || jobRoleRequest.getSpecification().isEmpty()) {
+            throw new InvalidJobRoleException("Specification cannot be empty");
+        }
+        if(jobRoleRequest.getSpecification().length() > 255){
```
Comment (jedgar3) [inline src/main/java/org/kainos/ea/validator/JobRoleValidator.java:30]: You have checked that specifictaion isn't more than 255 characters which is good, but you should also check if it is null because it is set to not null in the database

File: src/test/java/validator/JobRoleValidatorTest.java
Comment (jedgar3) [inline src/test/java/validator/JobRoleValidatorTest.java:1]: Add test for where specification is an empty string

**Review by jedgar3 — CHANGES_REQUESTED** [review]

File: src/main/java/org/kainos/ea/dao/BandDao.java (hunk 1 above)
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/dao/BandDao.java:59]: done

File: src/main/java/org/kainos/ea/model/JobRoleRequest.java (hunk 2 above)
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/model/JobRoleRequest.java:22]: done

File: src/main/java/org/kainos/ea/validator/JobRoleValidator.java
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/validator/JobRoleValidator.java:1]: thanks bro

File: src/test/java/integration/controller/JobRoleControllerIntegrationTest.java (hunk 3 above)
Comment (joshuasy10) [reply src/test/java/integration/controller/JobRoleControllerIntegrationTest.java:60]: done

File: src/test/java/validator/JobRoleValidatorTest.java
Comment (joshuasy10) [reply src/test/java/validator/JobRoleValidatorTest.java:1]: done

**Review by jedgar3 — APPROVED** [review]
Outsatnding work Josh

File: src/main/java/org/kainos/ea/controller/BandController.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        try {
+            return Response.ok(bandService.getBandById(id)).build();
+        } catch (FailedToGetBandException | DatabaseConnectionException e) {
+            System.err.println(e.getMessage());
+            return Response.serverError().build();
+        } catch (BandDoesNotExistException e) {
+            System.err.println(e.getMessage());
+            return Response.status(Response.Status.NOT_FOUND).build();
+        }
+    }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/controller/BandController.java:46]: not required

File: src/main/java/org/kainos/ea/controller/CapabilityController.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        try {
+            return Response.ok(capabilityService.getCapabilityById(id)).build();
+        } catch (DatabaseConnectionException | FailedToGetCapabilityException e) {
+            System.err.println((e.getMessage()));
+            return Response.serverError().build();
+        } catch (CapabilityDoesNotExistException e) {
+            System.err.println((e.getMessage()));
+            return Response.status(Response.Status.NOT_FOUND).build();
+        }
+    }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/controller/CapabilityController.java:45]: not required

File: src/main/java/org/kainos/ea/controller/JobRoleController.java [hunk 7]
```diff
@@ -47,4 +46,21 @@ public Response getJobRoleById(@PathParam("id") int id) {
...
         }
     }
+
+    @POST
+    @Path("/job-roles")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response createJobRole(JobRoleRequest jobRoleRequest) {
+        try {
+            return Response.ok(jobRoleService.createJobRole(jobRoleRequest)).build();
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/controller/JobRoleController.java:56]: should use created response code

File: src/main/java/org/kainos/ea/dao/BandDao.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class BandDao {
+    public List<Band> getAllBands() throws SQLException, DatabaseConnectionException {
+        Connection c = DatabaseConnector.getConnection();
+
+        String selectStatement = "SELECT * FROM bands;";
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/dao/BandDao.java:16]: don't select *

File: src/main/java/org/kainos/ea/dao/BandDao.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+                    rs.getInt("id"),
+                    rs.getString("name"),
+                    rs.getString("level"),
+                    rs.getString("responsibilities")
+            );
+        }
+
+        return null;
+    }
+}
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/dao/BandDao.java:59]: not required

File: src/main/java/org/kainos/ea/dao/CapabilityDao.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        while (rs.next()) {
+
+            return new Capability(
+                    rs.getInt("capabilities.id"),
+                    rs.getString("capabilities.name"),
+                    rs.getString("capabilities.description")
+            );
+        }
+        return null;
+    }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/dao/CapabilityDao.java:47]: not required

File: src/main/java/org/kainos/ea/exception/FailedToGetBandException.java [hunk 11]
```diff
@@ -0,0 +1,8 @@
+package org.kainos.ea.exception;
+
+public class FailedToGetBandException extends Exception {
+    @Override
+    public String getMessage() {
+        return "Failed to get band from the database.";
+    }
+}
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/exception/FailedToGetBandException.java:8]: not required

File: src/main/java/org/kainos/ea/exception/FailedToGetCapabilityException.java [hunk 12]
```diff
@@ -0,0 +1,9 @@
+package org.kainos.ea.exception;
+
+public class FailedToGetCapabilityException extends Throwable {
+    @Override
+    public String getMessage(){
+        return "Failed to get capability from the database";
+    }
+
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/exception/FailedToGetCapabilityException.java:8]: not required

File: src/main/java/org/kainos/ea/model/Band.java [hunk 13]
```diff
@@ -41,4 +36,20 @@ public String getLevel() {
...
         this.level = level;
     }
+
+    public String getResponsibilities() {
+        return responsibilities;
+    }
+
+    public void setResponsibilities(String responsibilities) {
+        this.responsibilities = responsibilities;
+    }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/model/Band.java:46]: not required

File: src/main/java/org/kainos/ea/service/BandService.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+                return band;
+            } else {
+                throw new BandDoesNotExistException();
+            }
+        } catch (SQLException e) {
+            System.err.println(e.getMessage());
+            throw new FailedToGetBandException();
+        }
+    }
+
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/service/BandService.java:42]: not required

File: src/main/java/org/kainos/ea/service/CapabilityService.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+            if (capability != null) {
+                return capability;
+            }
+            else{
+                throw new CapabilityDoesNotExistException();
+            }
+        } catch (SQLException e) {
+            System.err.println(e.getMessage());
+            throw new FailedToGetCapabilityException();
+        }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/service/CapabilityService.java:37]: not required

File: src/test/java/integration/controller/CapabilityControllerIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Test
+    void getCapabilityById_shouldReturn404_whenIDNotFound() {
+        Response responseCapability = APP.client().target("http://localhost:8080/api/capabilities/" + 0)
+                .request()
+                .get(Response.class);
+
+        Assertions.assertEquals(404, responseCapability.getStatus());
+    }
+}
```
Comment (shaunganley) [inline src/test/java/integration/controller/CapabilityControllerIntegrationTest.java:50]: not required

File: src/test/java/integration/controller/JobRoleControllerIntegrationTest.java [hunk 17]
```diff
@@ -48,4 +55,13 @@ void getJobRoleById_shouldReturn400_whenIDNotFound() {
 
         Assertions.assertEquals(400, responseJobRole.getStatus());
     }
-}
+
+    @Test
+    void createJobRole_shouldReturn200_whenCreated() {
```
Comment (shaunganley) [inline src/test/java/integration/controller/JobRoleControllerIntegrationTest.java:60]: add a test with invalid data and expect a 400

File: src/test/java/unit/controller/CapabilityControllerUnitTest.java [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+        System.out.println(response.getStatus());
+        assert(500 == response.getStatus());
+    }
+
+    @Test
+    void getCapabilityById_shouldReturnCapability_whenServiceReturnsCapability() throws DatabaseConnectionException, FailedToGetCapabilityException, CapabilityDoesNotExistException {
+        Mockito.when(capabilityService.getCapabilityById(1)).thenReturn(capability);
+        Response response= capabilityController.getCapabilityById(1);
+        assert(200 == response.getStatus());
+    }
```
Comment (shaunganley) [inline src/test/java/unit/controller/CapabilityControllerUnitTest.java:79]: not required

File: src/test/java/unit/service/CapabilityServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+        assertEquals(capability, capabilityService.getCapabilityById(1));
+    }
+
+    @Test
+    void getCapabilityById_shouldThrowCapabilityDoesNotExistException_whenDaoReturnsNull() throws SQLException, DatabaseConnectionException {
+        Mockito.when(capabilityDao.getCapabilityById(1)).thenReturn(null);
+
+        assertThrows(CapabilityDoesNotExistException.class, () -> capabilityService.getCapabilityById(1));
+    }
+}
```
Comment (shaunganley) [inline src/test/java/unit/service/CapabilityServiceTest.java:63]: not required

File: src/test/java/unit/service/JobRoleServiceTest.java [hunk 20]
```diff
@@ -75,4 +86,12 @@ void getJobRolesById_shouldThrowSQLException_whenDaoThrowsSQLException() throws
         Mockito.when(jobRoleDao.getJobRoleById(1)).thenThrow(SQLException.class);
         assertThrows(FailedToGetJobRoleException.class, () -> jobRoleService.getJobRolesById(1));
     }
-}
\ No newline at end of file
+
+    @Test
+    void createJobRole_shouldReturnId_whenDaoReturnsId() throws SQLException, DatabaseConnectionException, InvalidJobRoleException, FailedToCreateJobRoleException {
+        Mockito.when(bandDao.getBandById(1)).thenReturn(band);
+        Mockito.when(capabilityDao.getCapabilityById(1)).thenReturn(capability);
```
Comment (shaunganley) [inline src/test/java/unit/service/JobRoleServiceTest.java:93]: we shouldn't be mocking these daos as they're only passed to the validator. we should be mocking the validator and add tests for exceptions being thrown

**Review by shaunganley — CHANGES_REQUESTED** [review]

File: src/main/java/org/kainos/ea/dao/BandDao.java (hunk 9 above)
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/dao/BandDao.java:59]: yes it is, its required for validation for if the band exists, we discussed this in teams

File: src/main/java/org/kainos/ea/controller/BandController.java (hunk 5 above)
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/controller/BandController.java:46]: removed

File: src/main/java/org/kainos/ea/controller/JobRoleController.java (hunk 7 above)
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/controller/JobRoleController.java:56]: done

File: src/main/java/org/kainos/ea/dao/BandDao.java (hunk 8 above)
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/dao/BandDao.java:16]: done

File: src/test/java/unit/controller/CapabilityControllerUnitTest.java (hunk 18 above)
Comment (joshuasy10) [reply src/test/java/unit/controller/CapabilityControllerUnitTest.java:79]: removed

File: src/test/java/integration/controller/JobRoleControllerIntegrationTest.java (hunk 17 above)
Comment (joshuasy10) [reply src/test/java/integration/controller/JobRoleControllerIntegrationTest.java:60]: done

File: src/test/java/unit/service/CapabilityServiceTest.java (hunk 19 above)
Comment (joshuasy10) [reply src/test/java/unit/service/CapabilityServiceTest.java:63]: removed

File: src/test/java/unit/service/JobRoleServiceTest.java (hunk 20 above)
Comment (joshuasy10) [reply src/test/java/unit/service/JobRoleServiceTest.java:93]: will this not mean i'd have to completely refactor everything to pass validator objects in contsructors?

Comment (shaunganley) [reply src/test/java/unit/service/JobRoleServiceTest.java:95]: yes, otherwise you can't mock out the validator which means your unit tests are testing the validator class as well so aren't actual unit tests

File: src/main/java/org/kainos/ea/dao/CapabilityDao.java (hunk 10 above)
Comment (shaunganley) [reply src/main/java/org/kainos/ea/dao/CapabilityDao.java:47]: happy that this is required in dao but not service

File: src/main/java/org/kainos/ea/dao/BandDao.java (hunk 9 above)
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/dao/BandDao.java]: i have removed the service methods and only have Dao

File: src/test/java/integration/controller/CapabilityControllerIntegrationTest.java (hunk 16 above)
Comment (joshuasy10) [reply src/test/java/integration/controller/CapabilityControllerIntegrationTest.java]: done

File: src/test/java/unit/service/JobRoleServiceTest.java (hunk 20 above)
Comment (joshuasy10) [reply src/test/java/unit/service/JobRoleServiceTest.java]: done
